go/store/nbs/table_index.go

```diff
@@ -0,0 +1,506 @@
+)
+
+type tableIndex interface {
+	// ChunkCount returns the total number of chunks in the indexed file.
+	ChunkCount() uint32
+	// EntrySuffixMatches returns true if the entry at index |idx| matches
+	// the suffix of the address |h|. Used by |Lookup| after finding
+	// matching indexes based on |Prefixes|.
+	EntrySuffixMatches(idx uint32, h *addr) (bool, error)
+	// IndexEntry returns the |indexEntry| at |idx|. Optionally puts the
+	// full address of that entry in |a| if |a| is not |nil|.
+	IndexEntry(idx uint32, a *addr) (indexEntry, error)
+	// Lookup returns an |indexEntry| for the chunk corresponding to the
+	// provided address |h|. Second returns is |true| if an entry exists
+	// and |false| otherwise.
+	Lookup(h *addr) (indexEntry, bool, error)
+	// Ordinals returns a slice of indexes which maps the |i|th chunk in
+	// the indexed file to its corresponding entry in index. The |i|th
+	// entry in the result is the |i|th chunk in the indexed file, and its
+	// corresponding value in the slice is the index entry that maps to it.
+	Ordinals() ([]uint32, error)
+	// Prefixes returns the sorted slice of |uint64| |addr| prefixes; each
+	// entry corresponds to an indexed chunk address.
+	Prefixes() ([]uint64, error)
+	// TableFileSize returns the total size of the indexed table file, in bytes.
+	TableFileSize() uint64
+	// TotalUncompressedData returns the total uncompressed data size of
+	// the table file. Used for informational statistics only.
+	TotalUncompressedData() uint64
+
+	// Close releases any resources used by this tableIndex.
+	Close() error
+
+	// Clone returns a |tableIndex| with the same contents which can be
+	// |Close|d independently.
+	Clone() (tableIndex, error)
+}
+
+func ReadTableFooter(rd io.ReadSeeker) (chunkCount uint32, totalUncompressedData uint64, err error) {
+	footerSize := int64(magicNumberSize + uint64Size + uint32Size)
+	_, err = rd.Seek(-footerSize, io.SeekEnd)
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	footer, err := iohelp.ReadNBytes(rd, int(footerSize))
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if string(footer[uint32Size+uint64Size:]) != magicNumber {
+		return 0, 0, ErrInvalidTableFile
+	}
+
+	chunkCount = binary.BigEndian.Uint32(footer)
+	totalUncompressedData = binary.BigEndian.Uint64(footer[uint32Size:])
+
+	return
+}
+
+// parses a valid nbs tableIndex from a byte stream. |buff| must end with an NBS index
+// and footer, though it may contain an unspecified number of bytes before that data.
+// |tableIndex| doesn't keep alive any references to |buff|.
+// Does not allocate new memory except for offsets, computes on buff in place.
+func parseTableIndex(buff []byte) (onHeapTableIndex, error) {
+	chunkCount, totalUncompressedData, err := ReadTableFooter(bytes.NewReader(buff))
+	if err != nil {
+		return onHeapTableIndex{}, err
+	}
+
+	iS := indexSize(chunkCount)
+	buff = buff[:len(buff)-footerSize]
+	// Trim away any extra bytes
+	buff = buff[uint64(len(buff))-iS:]
```

Comment:
Can we add a test that we don't actually do this in the case of mmapTableReader here?

The backing array is going to be retained in these cases, and we would rather not waste overallocated bytes...the caller should have no reason to give us a bunch of bytes that aren't part of the index here.

Reply:
@reltuk I changed it to throw an error if the buff was the wrong size and broke a bunch of tests. Traced one of them back to here:
https://github.com/dolthub/dolt/blob/76bff36c632e15980b73f99ec1262a4031b79106/go/store/nbs/file_table_persister.go#L93

In this case, do I copy the data needed data to a new slice to ensure we don't retain unnecessary memory?

Reply:
Yeah, make fsTablePersister copy it before it passes it in.

Reply:
I created a new function called parseTableIndexByCopy and updated functions that were failing tests. parseTableIndex by default will now throw an error if the length or capacity of the given buff isn't correct.

Commits are organized to highlight the non-testing changes